### PR TITLE
LYS - Disable the save changes button until changes are made

### LINKED ISSUE #47304: Disable save button until changes are made
On the Site Visibility settings tab, we can improve UX by disabling the save button until the user makes a change. Once saved, the button should revert back to disabled.

As a follow-up, it would be ideal to then apply this behavior to all settings tabs.

## Before

Image

## After

Image

### PR DESCRIPTION
**Pull request by moon0326**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #47304 

This PR disables the `Save changes` button until changes are made.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a new JN site with this branch.
2. Go to `WooCommerce -> Settings -> Site Visibility`
3. Confirm the `Save changes` button is in disabled state.
4. Make changes.
5. Confirm the button changes to enabled.
6. Make changes back to the original.
7. Confirm the button is back to disabled state.

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

LYS: disables the "Save changes" button until changes are made.

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/api/plugins.php [hunk 1]
```diff
@@ -111,6 +110,7 @@ public function test_activate_plugin() {
 		$response       = $this->server->dispatch( $request );
 		$data           = $response->get_data();
 		$active_plugins = Plugins::get_active_plugins();
+		var_dump($data);
```
Comment (psealock) [inline plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/api/plugins.php]: 🤭

Comment (moon0326) [reply plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/api/plugins.php]: @psealock Thank you for catching it 😄 Removed it in [Remove test code](https://github.com/woocommerce/woocommerce/pull/47316/commits/59d3c2a33164a63088459c2e124ad70f1c76dd79) 🙏

File: plugins/woocommerce-admin/client/launch-your-store/settings/slotfill.js [hunk 2]
```diff
@@ -42,13 +44,32 @@ const SiteVisibility = () => {
...
-			?.woocommerce_private_link
+			?.woocommerce_private_link || 'no'
 	);
 
+	useEffect( () => {
+		const setting = window?.wcSettings?.admin?.siteVisibilitySettings || {};
+		const initValues = {
+			comingSoon: setting.woocommerce_coming_soon,
+			storePagesOnly:
+				setting.woocommerce_store_pages_only === false ? 'no' : 'yes',
```
Comment (rjchow) [inline plugins/woocommerce-admin/client/launch-your-store/settings/slotfill.js]: I think this checking for `=== false` is causing the save button to be disabled only when all the values are 'yes', so it's not actually disabling the save button for any other combination of values

**Review by rjchow — CHANGES_REQUESTED** [review]

File: plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/api/plugins.php [hunk 3]
```diff
@@ -101,8 +101,10 @@ public function test_install_invalid_plugins_param() {
 	 */
 	public function test_activate_plugin() {
 		wp_set_current_user( $this->user );
-
 		$request = new WP_REST_Request( 'POST', $this->endpoint . '/activate' );
+		if ( ! is_plugin_active( 'woocommerce/woocommerce.php' ) ) {
```
Comment (rjchow) [inline plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/api/plugins.php]: is this for fixing this test failure? https://github.com/woocommerce/woocommerce/pull/47337

Comment (moon0326) [reply plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/api/plugins.php]: @rjchow Yeah, good find! I missed that. I'll rebase instead 👍

File: plugins/woocommerce-admin/client/launch-your-store/settings/slotfill.js (hunk 2 above)
Comment (moon0326) [reply plugins/woocommerce-admin/client/launch-your-store/settings/slotfill.js]: @rjchow Thanks! I'll re-test it.

Comment (moon0326) [reply plugins/woocommerce-admin/client/launch-your-store/settings/slotfill.js]: @rjchow  Fixed in [Fix incorrect comparison](https://github.com/woocommerce/woocommerce/pull/47316/commits/edbceef274b3b4ee4693c076ba827a3e1f38b978)

**Review by rjchow — APPROVED** [review]
lgtm, thanks!

File: plugins/woocommerce-admin/client/launch-your-store/settings/slotfill.js [hunk 4]
```diff
@@ -11,13 +11,15 @@ import {
...
 	createElement,
+	useEffect,
 } from '@wordpress/element';
 import { registerPlugin } from '@wordpress/plugins';
 import { __ } from '@wordpress/i18n';
 import classNames from 'classnames';
 import { useCopyToClipboard } from '@wordpress/compose';
 import { recordEvent } from '@woocommerce/tracks';
 import { getSetting } from '@woocommerce/settings';
+import { isEqual } from 'lodash';
```
Comment (chihsuan) [inline plugins/woocommerce-admin/client/launch-your-store/settings/slotfill.js]: We can replace `isEqual` with [fastDeepEqual](https://href.li/?https://www.npmjs.com/package/fast-deep-equal) so we don't have to add more lodash functions to our codebase (pdToLP-yS-p2).

**Review by chihsuan — COMMENTED** [review]
Tested well. Left a comment to avoid adding more lodash functions.


Also, I agree with @psealock that the [Save button behavior should be consistent across all settings screens](https://github.com/woocommerce/woocommerce/issues/47305#issuecomment-2103375135
). I think we should have a solution that works for all settings screens, not just this one. But that can be a separate task.

**Comment by adrianduffell** [comment]
> Also, I agree with @psealock that the [Save button behavior should be consistent across all settings screens](https://github.com/woocommerce/woocommerce/issues/47305#issuecomment-2103375135). I think we should have a solution that works for all settings screens, not just this one. But that can be a separate task.

I've now created a follow-up issue for this https://github.com/woocommerce/woocommerce/issues/47393.

File: plugins/woocommerce-admin/client/launch-your-store/settings/slotfill.js (hunk 4 above)
Comment (moon0326) [reply plugins/woocommerce-admin/client/launch-your-store/settings/slotfill.js]: @chihsuan TIL! Replaced it with simple === instead in https://github.com/woocommerce/woocommerce/pull/47316/commits/6e72292f72acc68fb1e1ea1a007742c873b0a463

**Pull request merged**